### PR TITLE
Add Open Telemetry Spans throughout the codebase

### PR DESCRIPTION
**Pull request by jpinz**
This pull request adds the OpenTelemetry library and introduces tracing capabilities to various files in the `updater` directory. The most important changes include adding OpenTelemetry tracing to the `api_client.rb` file, modifying the `configure` method in `opentelemetry.rb` to record update job errors and exceptions, and adding a new method `capture_exception` to the `service.rb` file.

Main OpenTelemetry changes:

* <a href="diffhunk://#diff-2bd2ce9445c04b07391fccad568e355e5e15ef5516d842efd887a0357c8f8410R5-R20">`updater/lib/dependabot/opentelemetry.rb`</a>: Added OpenTelemetry library, enabled OpenTelemetry, and modified the `configure` method for recording update job errors and exceptions. <a href="diffhunk://#diff-2bd2ce9445c04b07391fccad568e355e5e15ef5516d842efd887a0357c8f8410R5-R20">[1]</a> <a href="diffhunk://#diff-2bd2ce9445c04b07391fccad568e355e5e15ef5516d842efd887a0357c8f8410L19-R31">[2]</a> <a href="diffhunk://#diff-2bd2ce9445c04b07391fccad568e355e5e15ef5516d842efd887a0357c8f8410R43-R94">[3]</a>
* <a href="diffhunk://#diff-1f8c3901647148d6f9fbda3a5c943490beb28c1291f21a3b6713471e0c39902bR50-R60">`updater/lib/dependabot/api_client.rb`</a>: Added OpenTelemetry tracing to various methods in the `api_client.rb` file, creating new spans and setting attributes for job-related information. <a href="diffhunk://#diff-1f8c3901647148d6f9fbda3a5c943490beb28c1291f21a3b6713471e0c39902bR50-R60">[1]</a> <a href="diffhunk://#diff-1f8c3901647148d6f9fbda3a5c943490beb28c1291f21a3b6713471e0c39902bR210-R216">[2]</a> <a href="diffhunk://#diff-1f8c3901647148d6f9fbda3a5c943490beb28c1291f21a3b6713471e0c39902bR166-R171">[3]</a> <a href="diffhunk://#diff-1f8c3901647148d6f9fbda3a5c943490beb28c1291f21a3b6713471e0c39902bR150-R153">[4]</a> <a href="diffhunk://#diff-1f8c3901647148d6f9fbda3a5c943490beb28c1291f21a3b6713471e0c39902bR127-R128">[5]</a> <a href="diffhunk://#diff-1f8c3901647148d6f9fbda3a5c943490beb28c1291f21a3b6713471e0c39902bR189">[6]</a> <a href="diffhunk://#diff-1f8c3901647148d6f9fbda3a5c943490beb28c1291f21a3b6713471e0c39902bR230">[7]</a> <a href="diffhunk://#diff-1f8c3901647148d6f9fbda3a5c943490beb28c1291f21a3b6713471e0c39902bR33-R37">[8]</a> <a href="diffhunk://#diff-1f8c3901647148d6f9fbda3a5c943490beb28c1291f21a3b6713471e0c39902bR99-R105">[9]</a> <a href="diffhunk://#diff-1f8c3901647148d6f9fbda3a5c943490beb28c1291f21a3b6713471e0c39902bR79-R86">[10]</a>
* <a href="diffhunk://#diff-b4b880c69008ab386f67bd253c4cbfc68fae8de2aeb5dfad140208f278f1008fR16">`updater/lib/dependabot/update_files_command.rb`</a>: Added OpenTelemetry tracing to the `perform_job` method, creating a new span. <a href="diffhunk://#diff-b4b880c69008ab386f67bd253c4cbfc68fae8de2aeb5dfad140208f278f1008fR16">[1]</a> <a href="diffhunk://#diff-b4b880c69008ab386f67bd253c4cbfc68fae8de2aeb5dfad140208f278f1008fR47">[2]</a>
* <a href="diffhunk://#diff-4c01bde1823e41430372a425ad88fc23267edb4c723890ec6dee64d9581a64c4R19">`updater/lib/dependabot/file_fetcher_command.rb`</a>: Added OpenTelemetry tracing to the `perform_job` method, creating a new span. <a href="diffhunk://#diff-4c01bde1823e41430372a425ad88fc23267edb4c723890ec6dee64d9581a64c4R19">[1]</a> <a href="diffhunk://#diff-4c01bde1823e41430372a425ad88fc23267edb4c723890ec6dee64d9581a64c4R54">[2]</a>

Additional changes:

* <a href="diffhunk://#diff-72d65f40e823f48b30f91700de1178de9a24ad4530b76aab448c7342ea56d494L92-R101">`updater/lib/dependabot/service.rb`</a>: Added a new method `capture_exception` to the `service.rb` file, which calls the `record_exception` method from the `Dependabot::OpenTelemetry` module and captures the exception using the Raven gem.

### CODE REVIEW

**Comment by JamieMagee** [comment]
For anyone reviewing this PR, it's easiest if you ignore whitespace changes.

File: updater/lib/dependabot/opentelemetry.rb [hunk 1]
```diff
@@ -15,19 +16,69 @@ def self.should_configure?
...
+        attributes.store("dependabot.job.error_details.#{key}", value)
+      end
+
+      current_span.add_event(error_type, attributes: attributes)
+    end
+
+    sig do
+      params(
+      error: StandardError,
+      job: T.untyped,
```
Comment (JamieMagee) [inline updater/lib/dependabot/opentelemetry.rb:69]: Can this be be
```suggestion
      job: Dependabot::Job,
```

Or another type?

File: updater/lib/dependabot/opentelemetry.rb [hunk 2]
```diff
@@ -15,19 +16,69 @@ def self.should_configure?
...
         config.use "OpenTelemetry::Instrumentation::Http"
       end
+
+      tracer
+    end
+
+    sig { returns(T.nilable(::OpenTelemetry::Trace::Tracer)) }
+    def self.tracer
+      return unless should_configure?
+      return ::OpenTelemetry.tracer_provider.tracer('dependabot')
```
Comment (JamieMagee) [inline updater/lib/dependabot/opentelemetry.rb:44]: This takes an optional version parameter. 
```suggestion
      return ::OpenTelemetry.tracer_provider.tracer('dependabot', Dependabot::VERSION)
```

File: updater/lib/dependabot/opentelemetry.rb [hunk 3]
```diff
@@ -15,19 +16,69 @@ def self.should_configure?
     sig { void }
     def self.configure
```
Comment (JamieMagee) [inline updater/lib/dependabot/opentelemetry.rb:17]: We talked about this offline, but we might want to consider calling `start_root_span` in `configure` and then using `current_span` as a less invasive way to try this out.

File: updater/lib/dependabot/api_client.rb [hunk 4]
```diff
@@ -125,37 +132,44 @@ def record_update_job_unknown_error(error_type:, error_details:)
...
-    rescue HTTP::ConnectionError, OpenSSL::SSL::SSLError
-      retry_count ||= 0
-      retry_count += 1
-      raise if retry_count > 3
-
-      sleep(rand(3.0..10.0))
-      retry
+      ::Dependabot::OpenTelemetry.tracer.in_span("mark_job_as_processed", kind: :internal) do |span|
+        span.set_attribute("base_commit_sha", base_commit_sha)
+        span.set_attribute("dependabot.job.id", job_id)
```
Comment (JamieMagee) [inline updater/lib/dependabot/api_client.rb:137]: We should extract all these attribute names to a constants file somewhere, to make them easier to find and reference.

File: updater/lib/dependabot/opentelemetry.rb (hunk 1 above)
Comment (jpinz) [reply updater/lib/dependabot/opentelemetry.rb]: I switched it, appears to work fine

File: updater/lib/dependabot/opentelemetry.rb [hunk 5]
```diff
@@ -15,19 +16,69 @@ def self.should_configure?
...
       require "opentelemetry/instrumentation/http"
+      ENV['OTEL_TRACES_EXPORTER'] ||= 'console'
 
       ::OpenTelemetry::SDK.configure do |config|
         config.service_name = "dependabot"
+        config.add_span_processor(
+              ::OpenTelemetry::SDK::Trace::Export::SimpleSpanProcessor.new(
+                ::OpenTelemetry::SDK::Trace::Export::ConsoleSpanExporter.new
+              )
+            )
```
[the remark this answers is not in the document]
Comment (jpinz) [reply updater/lib/dependabot/opentelemetry.rb]: Removing this while it's set to export to console makes it not work. Will have to investigate further.

[the remark this answers is not in the document]
Comment (jpinz) [reply updater/lib/dependabot/opentelemetry.rb]: I think I got it working

File: updater/lib/dependabot/opentelemetry.rb (hunk 2 above)
Comment (jpinz) [reply updater/lib/dependabot/opentelemetry.rb]: Done

File: updater/lib/dependabot/opentelemetry.rb (hunk 3 above)
Comment (jpinz) [reply updater/lib/dependabot/opentelemetry.rb:25]: Is there any reason for me to refactor it from how I currently implemented it?

Comment (JamieMagee) [reply updater/lib/dependabot/opentelemetry.rb:27]: I thought we could make the changes less invasive if we didn't wrap everything in blocks, but it's mostly a stylistic change.

Comment (jpinz) [reply updater/lib/dependabot/opentelemetry.rb:27]: I think I'd rather have the name of the span be the method name. Vs one root one to use everywhere

**Pull request closed without merging**